xline/src/server/lease_server.rs

```diff
@@ -346,4 +298,70 @@ where
         };
         Ok(tonic::Response::new(res))
     }
+
+    /// LeaseRevoke revokes a lease. All keys attached to the lease will expire and be deleted.
+    async fn lease_revoke(
+        &self,
+        request: tonic::Request<LeaseRevokeRequest>,
+    ) -> Result<tonic::Response<LeaseRevokeResponse>, tonic::Status> {
+        debug!("Receive LeaseRevokeRequest {:?}", request);
+
+        let is_fast_path = true;
+        let (res, sync_res) = self.propose(request, is_fast_path).await?;
+
+        let mut res: LeaseRevokeResponse = res.decode().into();
+        if let Some(sync_res) = sync_res {
+            let revision = sync_res.revision();
+            debug!("Get revision {:?} for LeaseRevokeResponse", revision);
+            if let Some(mut header) = res.header.as_mut() {
+                header.revision = revision;
+            }
+        }
+        Ok(tonic::Response::new(res))
+    }
+
+    /// LeaseTimeToLive retrieves lease information.
+    async fn lease_time_to_live(
+        &self,
+        request: tonic::Request<LeaseTimeToLiveRequest>,
+    ) -> Result<tonic::Response<LeaseTimeToLiveResponse>, tonic::Status> {
+        debug!("Receive LeaseTimeToLiveRequest {:?}", request);
+        loop {
+            if self.lease_storage.is_primary() {
+                // TODO wait applied index
+                let time_to_live_req = request.into_inner();
+                let Some(lease) = self.lease_storage.look_up(time_to_live_req.id) else {
+                    return Err(tonic::Status::not_found("Lease not found"));
+                };
+
+                let keys = time_to_live_req
+                    .keys
+                    .then(|| lease.keys())
+                    .unwrap_or_default();
+                let res = LeaseTimeToLiveResponse {
+                    header: Some(self.lease_storage.gen_header()),
+                    id: time_to_live_req.id,
+                    ttl: lease.remaining().as_secs().cast(),
+                    granted_ttl: lease.ttl().as_secs().cast(),
+                    keys,
+                };
+                return Ok(tonic::Response::new(res));
+            }
+            let leader_id = self.client.get_leader_id().await;
+            let leader_addr = self.all_members.get(&leader_id).unwrap_or_else(|| {
+                unreachable!(
+                    "The address of leader {} not found in all_members {:?}",
+                    leader_id, self.all_members
+                )
+            });
+            if !self.lease_storage.is_primary() {
+                let mut lease_client = LeaseClient::connect(format!("http://{leader_addr}"))
+                    .await
+                    .map_err(|e| {
+                        tonic::Status::internal(format!("Connect to leader error: {e}"))
+                    })?;
+                return lease_client.lease_time_to_live(request).await;
+            }
+        }
```

Comment:
Let's assume that the current server is a candidate and the lease store it has is not a primary lessor. So the self.client will wait for the new leader id to return. If the candidate lost the election, it should redirect this request to the real leader, otherwise, it should deal with this request by itself.